### DISCUSSION TITLE
NAT Rules on Initiator Not Working

### DISCUSSION THREAD

**Discussion by blueaxions** [strongswan/strongswan#1648] [General]
Hello,

Running Strongswan on Debian Bullseye system which initiates VPN sessions but would like to make this system into a VPN router to be used in hotel hotspots when traveling where different devices can connect to it.

To do this, I'm adding the following NAT rules and it's failing with this error "iptables v1.8.7 (legacy): Couldn't load match `policy':No such file or directory"

iptables -t nat -A POSTROUTING -s 10.1.0/24 -o eth0 -m policy --dir out --pol ipsec -j ACCEPT
iptables -t nat -A POSTROUTING -s 10.1.0/24 -o eth0 -j MASQUERADE

Any help is greatly appreciated!

-SR

**Comment by tobiasbrunner** [comment]
Could be an issue with the kernel and its modules or something related to nft vs. iptables. I think the policy match requires the `xt_policy` module.

**Comment by blueaxions** [comment]
I figured as much, so I cross-compiled Debian Bullseye in the build environment and tried installing just the xt_policy.ko file on the target system. Next I ran the same 'iptables' command as before and it ran without any errors but it appears that the command didn't add the necessary rule. Below is the output:

iptables -L -v -n
Chain INPUT (policy ACCEPT 4675 packets, 524K bytes)
pkts bytes target     prot opt in     out     source               destination

Chain FORWARD (policy ACCEPT 5 packets, 316 bytes)
pkts bytes target     prot opt in     out     source               destination

Chain OUTPUT (policy ACCEPT 4668 packets, 523K bytes)
pkts bytes target     prot opt in     out     source               destination

Shouldn't this be taken care of by the _updown script?

-SR

**Reply by tobiasbrunner** [reply]
> iptables -L -v -n

This will not show rules in the `nat` table. Add `-t nat` for that, or even better, use `iptables-save` to see all rules.

> Shouldn't this be taken care of by the _updown script?

What do you mean? Did you configure a custom script?

**Comment by blueaxions** [comment]
Here's the output of iptables-save:

iptables-save
# Generated by iptables-save v1.8.7 on Wed Apr 12 10:04:55 2023
*filter
:INPUT ACCEPT [7937:891233]
:FORWARD ACCEPT [5:316]
:OUTPUT ACCEPT [7929:890430]
COMMIT
# Completed on Wed Apr 12 10:04:55 2023
# Generated by iptables-save v1.8.7 on Wed Apr 12 10:04:55 2023
*nat
:PREROUTING ACCEPT [11:658]
:INPUT ACCEPT [1:96]
:OUTPUT ACCEPT [190:14440]
:POSTROUTING ACCEPT [190:14440]
-A POSTROUTING -s 10.10.10.0/24 -o wlan0 -m policy --dir out --pol ipsec -j ACCEPT
-A POSTROUTING -s 10.10.10.0/24 -o wlan0 -m policy --dir out --pol ipsec -j ACCEPT
-A POSTROUTING -s 10.10.10.0/24 -o wlan0 -j MASQUERADE
-A POSTROUTING -s 10.10.10.0/24 -o wlan0 -m policy --dir out --pol ipsec -j ACCEPT
-A POSTROUTING -s 10.10.10.0/24 -o wlan0 -j MASQUERADE
-A POSTROUTING -s 10.10.10.0/24 -o wlan0 -m policy --dir out --pol ipsec -j ACCEPT
COMMIT
# Completed on Wed Apr 12 10:04:55 2023

Regarding the _updown script, I was referring to the one that comes with the strongSwan package. At least on the Responder running on RHEL 8, I didn't have to create my own _updown script, instead I referenced it in the swanctl.conf file and it auto-magically loads and clears the rules when tunnels are up or down.

-SR

**Reply by tobiasbrunner** [reply]
There is a [default updown script](https://docs.strongswan.org/docs/5.9/plugins/updown.html), but that does not install any NAT rules. It only installs rules in FORWARD and INPUT/OUTPUT to allow traffic if the default policies for these chains is DROP or REJECT. You either have to extend that script or just write a custom one to install NAT rules dynamically (e.g. if you want to NAT to a virtual IP address that's assigned by the responder, but depending on the use case installing them statically might also work).

**Reply by blueaxions** [reply]
Got it! I'll look into modifying the default script to add the NAT rules. 
From the output I posted, do the NAT rules look correct to you? Or am I missing anything?

-SR

**Reply by tobiasbrunner** [reply]
Depends on your use case and the negotiated traffic selectors (e.g. if a virtual IP is used or you protect the physical IP of the VPN gateway) as with `MASQUERADE` you don't really have control over the IP to which traffic is mapped. And the `ACCEPT` rule is probably not necessary. [Here is an example updown script](https://github.com/strongswan/strongswan/blob/master/testing/tests/ikev2/nat-virtual-ip/hosts/moon/etc/nat_updown) that installs NAT rules that map to a virtual IP.

**Comment by blueaxions** [comment]
Thanks, I'll implement the updown script and will post the results. 

What we're trying to solve is for a road warrior Windows machine to be able to ping a system that's behind two strongSwan gateways and its not able to. One strongSwan gateway runs Debian and the other runs RHEL 8 and the tunnel is established between the two gateways. From the gateway running Debian I'm able to ping the target system through the tunnel. The gateway running Debian is where I have the NAT and MASQUERADE rules in place.

The road warrior Windows system is connected to the gateway running Debian and I'm unable to ping the target system from the Windows system. Running tcpdump on the gateway running RHEL 8 I noticed packets from the Windows machine arriving without NATing even though the Debian gateway has the iptables NAT rules in place.

-SR

**Reply by tobiasbrunner** [reply]
If you need help with that, you need to post a log more information about all of this (configs, status output, firewall rules etc. basically see #196).
